### PR TITLE
Fix infinite loop case

### LINKED ISSUE #2: An infinite loop with unresolved collisions
Hi,

I have found another error. In some cases, when code gets some unresolved collisions, it freezes and stuck in an infinite loop. Here is the package with an example: 
[infloop.zip](https://github.com/tolcz/ABCDeGraphGenerator.jl/files/6955170/infloop.zip)

### PR DESCRIPTION
**Pull request by bkamins**
@bartoszpankratz + @tolcz

this should fix https://github.com/tolcz/ABCDeGraphGenerator.jl/issues/2

can you please test (I do not have access to computer where I can run Julia currently)? Thank you!

### CODE REVIEW

File: src/graph_sampler.jl [hunk 1]
```diff
@@ -318,8 +329,45 @@ function config_model(clusters, params)
     old_len = length(edges)
     union!(edges, global_edges)
     @assert length(edges) == old_len + length(global_edges)
-    @assert 2 * length(global_edges) == length(stubs)
-    edges
+
```
Comment (bkamins) [inline src/graph_sampler.jl]:
```suggestion
    if isempty(recycle)
        @assert 2 * length(global_edges) == length(stubs)
    else
```

File: src/graph_sampler.jl [hunk 2]
```diff
@@ -318,8 +329,45 @@ function config_model(clusters, params)
...
+            if (newp[1] == newp[2]) || (newp in global_edges) || (newp in edges)
+                push!(recycle, newp)
+            else
+                push!(edges, newp)
+            end
+        end
+    end
+    if isempty(recycle)
+        @assert 2 * length(global_edges) == length(stubs)
+    else
```
Comment (bkamins) [inline src/graph_sampler.jl]:
```suggestion
    if !isempty(recycle)
```

File: src/graph_sampler.jl [hunk 3]
```diff
@@ -318,8 +329,45 @@ function config_model(clusters, params)
...
+        p2 = pop!(edges, x)
+        if rand() < 0.5
+            newp1 = minmax(p1[1], p2[1])
+            newp2 = minmax(p1[2], p2[2])
+        else
+            newp1 = minmax(p1[1], p2[2])
+            newp2 = minmax(p1[2], p2[1])
+        end
+        for newp in (newp1, newp2)
+            if (newp[1] == newp[2]) || (newp in global_edges) || (newp in edges)
```
Comment (bkamins) [inline src/graph_sampler.jl]:
```suggestion
            if (newp[1] == newp[2]) || (newp in edges)
```

**Comment by tlcz** [comment]
Hi @bkamins 
branch `bkamins-patch-1` fails to compile on my env:
```julia
julia> using ABCDGraphGenerator
[ Info: Precompiling ABCDGraphGenerator [4c9194b5-d5d9-4bea-bd96-43ca8ce89b0b]
ERROR: LoadError: LoadError: syntax: incomplete: "function" at ABCDGraphGenerator.jl/src/graph_sampler.jl:140 requires end
Stacktrace:
  [1] top-level scope
    @ ABCDGraphGenerator.jl/src/graph_sampler.jl:140
  [2] include(mod::Module, _path::String)
    @ Base ./Base.jl:386
  [3] include(x::String)
    @ ABCDGraphGenerator ABCDGraphGenerator.jl/src/ABCDGraphGenerator.jl:1
  [4] top-level scope
    @ ABCDGraphGenerator.jl/src/ABCDGraphGenerator.jl:7
  [5] include
    @ ./Base.jl:386 [inlined]
  [6] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt64}}, source::Nothing)
    @ Base ./loading.jl:1235
  [7] top-level scope
    @ none:1
  [8] eval
    @ ./boot.jl:360 [inlined]
  [9] eval(x::Expr)
    @ Base.MainInclude ./client.jl:446
 [10] top-level scope
    @ none:1
in expression starting at ABCDGraphGenerator.jl/src/graph_sampler.jl:140
in expression starting at ABCDGraphGenerator.jl/src/ABCDGraphGenerator.jl:1
```

File: src/graph_sampler.jl [hunk 4]
```diff
@@ -360,9 +362,7 @@ function config_model(clusters, params)
             end
         end
     end
-    if isempty(recycle)
-        @assert 2 * length(global_edges) == length(stubs)
-    else
+    if !isempty(recycle)
```
Comment (tolcz) [inline src/graph_sampler.jl:365]: probably a missing `end` here

File: src/graph_sampler.jl [hunk 5]
```diff
@@ -329,7 +329,9 @@ function config_model(clusters, params)
     old_len = length(edges)
     union!(edges, global_edges)
     @assert length(edges) == old_len + length(global_edges)
-
+    if isempty(recycle)
+        @assert 2 * length(global_edges) == length(stubs)
+    else
```
Comment (tlcz) [inline src/graph_sampler.jl:334]: probably an `end` missing here

**Comment by tlcz** [comment]
After adding a missing `end` the 'infinite loop' problem seems resolved, at least for the attached configuration.

**Comment by tlcz** [comment]
A general remark is that now 90% of `config_model()`'s code seems dedicated to 'collision resolution' logic. It looks overcomplicated and should probably be redesigned and simplified.

**Comment by bkamins** [comment]
Thank you for testing and for the fix. I will wait with merging till we decide if we should not simplify collision resolution.

**Comment by bkamins** [comment]
@bartoszpankratz - is all working correctly for you in this branch (so that I can merge it to `master`)? Thank you!

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**